File: src/main/java/util/notepad/LinuxNotepad.java
package util.notepad;

import static util.extension.RobotExtension.screenHeight;
import static util.extension.RobotExtension.screenWidth;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.io.File;
import java.io.IOException;
import lombok.experimental.ExtensionMethod;
import lombok.extern.log4j.Log4j2;
import util.extension.RobotExtension;

@Log4j2
@ExtensionMethod({RobotExtension.class})
public class LinuxNotepad implements Notepad {
	public static final String home = System.getProperty("user.home") + File.separator + "ParseButPro";
	private Robot r;
	Process process;

	LinuxNotepad() {
		try {
			r = new Robot();
		} catch (AWTException e) {
			log.error("Failed to initialize Robot", e);
		}
	}

	public void openNotepad() throws IOException {
		ProcessBuilder processBuilder = new ProcessBuilder("notepadqq");
		process = processBuilder.start();
		r.delay(1000);
		r.mouseMove(screenWidth / 2, screenHeight / 2);
		r.leftClick();
	}

	public void writeText(String text) {
		r.type(text);
	}

	public void deleteText() {
		r.control(KeyEvent.VK_A);
		r.type(KeyEvent.VK_DELETE);
	}

	public void addNewLine() {
		r.enter();
	}

	public void saveFileAs(String name) {
		r.type(name);
		r.control(KeyEvent.VK_S);
		for (int i = 0; i < 7; i++) {
			r.type(KeyEvent.VK_TAB);
		}
		r.enter();
		r.type(home);
		r.enter();
	}

	public void openNewFile() {
		r.control(KeyEvent.VK_N);
	}

	public void closeFile() {
		r.control(KeyEvent.VK_W);
	}

	public void closeNotepad() throws InterruptedException {
		r.control(KeyEvent.VK_F4);
		log.info("Exited with code: {}", process.waitFor());
	}
}

File: src/main/java/util/notepad/LinuxNotepad.kt
package util.notepad

import org.apache.logging.log4j.LogManager
import util.extension.*
import java.awt.Robot
import java.awt.event.KeyEvent
import java.io.File
import java.io.IOException

class LinuxNotepad : Notepad {
  private val log = LogManager.getLogger()
  private val robot = Robot()
  private var process: Process? = null

  companion object {
    val homeDirectory = System.getProperty("user.home") + File.separator + "ParseButPro"
  }

  @Throws(IOException::class)
  override fun openNotepad() {
    process = ProcessBuilder("notepadqq").start().also {
      robot.apply {
        delay(1000)
        mouseMove(RobotUtils.screenWidth / 2, RobotUtils.screenHeight / 2)
        leftClick()
      }
    }
  }

  override fun writeText(text: String) {
    robot.type(text)
  }

  override fun deleteText() {
    robot.apply {
      control(KeyEvent.VK_A)
      type(KeyEvent.VK_DELETE)
    }
  }

  override fun addNewLine() {
    robot.enter()
  }

  override fun saveFileAs(name: String) {
    robot.apply {
      type(name)
      control(KeyEvent.VK_S)
      repeat(7) { type(KeyEvent.VK_TAB) }
      enter()
      type(homeDirectory)
      enter()
    }
  }

  override fun openNewFile() {
    robot.control(KeyEvent.VK_N)
  }

  override fun closeFile() {
    robot.control(KeyEvent.VK_W)
  }

  @Throws(InterruptedException::class)
  override fun closeNotepad() {
    robot.control(KeyEvent.VK_F4)
    process?.waitFor()?.also { log.debug("Exited Notepad++ with code: {}", it) }
  }
}